src/ASM/ASMs2Drecovery.C

```diff
@@ -246,19 +248,24 @@ bool ASMs2D::assembleL2matrices (SparseMatrix& A, StdVector& B,
 	    if (dJw == 0.0) continue; // skip singular points
 	  }
 
-	  // Integrate the linear system A*x=B
+          // Integrate the mass matrix
+          eA.outer_product(phi, phi, true, dJw);
+
+	  // Integrate the rhs vector B
 	  for (size_t ii = 0; ii < phi.size(); ii++)
 	  {
-	    int inod = MNPC[iel][ii]+1;
-	    for (size_t jj = 0; jj < phi.size(); jj++)
-	    {
-	      int jnod = MNPC[iel][jj]+1;
-	      A(inod,jnod) += phi[ii]*phi[jj]*dJw;
-	    }
 	    for (size_t r = 1; r <= sField.rows(); r++)
-	      B(inod+(r-1)*nnod) += phi[ii]*sField(r,ip+1)*dJw;
+	      eB(1+ii+(r-1)*p1*p2) += phi[ii]*sField(r,ip+1)*dJw;
```

Comment:
Yes, but I forgot to mention (except for in the previous review some time ago) that you then can use Vector::add instead for the loop (to be aligned with the outer_product usage for the element matrix). I.e. something like this should work:
```
Vectors eB(sField.rows(),Vector(p1*p2);
. . .
    // Integrate the rhs vector B
    for (size_r t = 1; r <= sField.rows(); r++)
      eB[r].add(phi,sField(r,ip+1)*dJw);
```